### ISSUE TITLE
Crash when vsync is off

### ISSUE THREAD

**Issue by exuvo** [Noxagonal/Vulkan2DRenderer#152]
If i turn off vsync i sometimes get a segmentation fault in libvulkan.so.1 when calling vk2d::Window::BeginRender. Not immediately but usually anywhere between 0-2 minutes after creating the window. I have a time controlled loop so that even with vsync off it keeps 60fps. If i turn off that time control and just let it render as fast as it can it crashes in the first second.

More specifically the crash is at `vk2d::_internal::WindowImpl::BeginRender:1345` `vkBeginCommandBuffer`. 

Not really a problem for me right now as i'll just keep vsync on for the time being.

**Comment by Noxagonal** [comment]
Just a hunch for now.
Command buffers must not be touched on CPU side until the GPU is finished with it, this would always crash.
I had some issues with `SynchronizeFrame()` seemingly not synchronizing in `ReCreateSwapchain():3252` so I added `vkQueueWaitIdle( primary_render_queue.GetQueue() );` on the next line as a patch.

I'll investigate `SynchronizeFrame()` soon.

**Comment by Noxagonal** [comment]
Can't reproduce on Windows, I'll try again on Linux later...

**Comment by exuvo** [comment]
Both my linux arch and my friends ubuntu crash with vsync off at the above mentioned place. next_image is not equal to previous_image so SynchronizeFrame is not run when it crashes. I have next_image 3 and previous_image 2. But vk_render_command_buffers is only of size 3 for me so it is going outside the array and sending invalid command buffer into begin.

**Comment by Noxagonal** [comment]
Okay. There should definitely not be any local variables with the same name as any of the class members anywhere in the code so this is my oversight.
Thanks for catching it. :)

TL;DR removing the line `uint32_t swapchain_image_count = 0;` is the correct fix. Should be fixed now on the latest master.

I don't know why that local variable is there but I was probably testing something and forgot it.

In detail if you're curious:
- I give the `swapchain_image_count` to `VkSwapchainCreateInfoKHR::minImageCount` with the minimum number of swapchain images to create.`vkCreateSwapchainKHR` only guarantees a minimum number of swapchain images, it can create more.
- When we call `vkGetSwapchainImagesKHR` we get the actual number of swapchain images that the OS is giving us. Windows seems to always create the minimum number so I didn't catch this earlier.
- The reason why it can take a while to crash is because `vkAcquireNextImageKHR` is allowed to return swapchain images out of order, whichever is available to be used first. Both Windows and Linux seem to favor lower indices.

**Comment by exuvo** [comment]
Yeah i eventually figured out what that part of the code was doing after discussing with axlar2 who has been doing a vulkan tutorial on the side to learn vulkan. I am getting much lower input latencies with mailbox so it was really nice to get this fixed (tested by dragging around imgui window).
